File: QuickSortPart2.c
#include <stdio.h>
#include <string.h>
#include <math.h>
#include <stdlib.h>
#include <assert.h>
void printArray(int *ar,int low,int high){
    int i=0;
    printf("\n");
    for(i=low;i<=high;i++){
        
        printf("%d ",ar[i]);
    }
}
void swap(int *a,int *b){
    int temp;
    temp=*a;
    *a=*b;
    *b=temp;
    
}
int partition(int *ar,int p,int r){
    
    int x=ar[p];
    int i=p,j;
    for(j=p+1;j<r;j++){
        if(ar[j]<=x){
            i++;
            swap(&ar[j],&ar[i]);
        }
        
    }
    swap(&ar[p],&ar[i]);
    return i;
}
void quickSort(int *  ar,int low,int high) {
    int pivot;
    if(low<high){
        pivot=partition(ar,low,high);
        quickSort(ar,low,pivot-1);
        printArray(ar,low,pivot-1);
        quickSort(ar,pivot+1,high);
        printArray(ar,pivot+1,high);
        
    }
    

}

int main(void) {
   
   int _ar_size;
scanf("%d", &_ar_size);
int _ar[_ar_size], _ar_i;
for(_ar_i = 0; _ar_i < _ar_size; _ar_i++) { 
   scanf("%d", &_ar[_ar_i]); 
}

quickSort(_ar,0,_ar_size);
   
   return 0;
}

File: QuickSortPart2.java

import java.util.*;
public class Solution {
       
          static void quickSort(int[] ar,int low,int high) {
              int pivotLocation;
              if(low<high){
                  pivotLocation=sortUsingQuick(ar,low,high);
                  //System.out.print("pivot "+pivotLocation);
				  //printArray(ar);
				  quickSort(ar,low,pivotLocation);
				  
                  /*for(int i=low;i<=pivotLocation-1;i++){
                      System.out.print(ar[i]+" ");
                      
                  }*/
                  //System.out.println("");
				 quickSort(ar,pivotLocation+1,high);
                  /*for(int i=pivotLocation+1;i<=high;i++){
                      System.out.print(ar[i]+" ");
                      
                  }*/
				 //System.out.print("low "+low+" high "+high);
                 printArray(ar,low,high); 
              }
			  
                    
       }
	   static int sortUsingQuick(int[] ar,int low,int high){
		List<Integer> smallerValues = new ArrayList<Integer>();
                List<Integer> biggerValues = new ArrayList<Integer>();
                int toPartitionAround = ar[low];
                for(int c=low+1;c<high;c++) {
                    if(ar[c] < toPartitionAround) {
                        smallerValues.add(ar[c]);
                    } else if(ar[c]>toPartitionAround){
                         biggerValues.add(ar[c]);
                    }
                }
      for(int c=0;c<smallerValues.size();c++) {
            ar[low+c] = smallerValues.get(c);
            
          
      }
      ar[smallerValues.size()+low] = toPartitionAround;
      for(int m=0;m<biggerValues.size();m++) {
          ar[low+m+ smallerValues.size() + 1] = biggerValues.get(m);
      }
           return low+smallerValues.size();
      //printArray(ar); 
	  
          }
 
 static void printArray(int[] ar,int low,int high) {
		if(low+1!=high){
         for(int i=low;i<high;i++){
            System.out.print(ar[i]+" ");
         }
           System.out.println("");
		   }
      }
       
      public static void main(String[] args) {
           Scanner in = new Scanner(System.in);
           int n = in.nextInt();
           int[] ar = new int[n];
           for(int i=0;i<n;i++){
              ar[i]=in.nextInt(); 
           }
           quickSort(ar,0,n);
       }    
   }
